### PR TITLE
Add `strict` option

### LINKED ISSUE #26: Throttle limit exceeded after a delay
I am calling an external service with a limit of 10 calls per second. I have configured the throttle as
`throttle = pThrottle({
                limit: 10,
                interval: 1010
            })`

Occasionally we are receiving a throttling error from the external service. Checking our logs, the service was called 11 times within a period of 980 ms.

The following mocha test reproduces the issue:
```javascript
'use strict';

const pThrottle = require('p-throttle');

require('should');

describe('p-throttle', () => {
    let throttle,
        throttled;

    beforeEach(() => {
        throttle = pThrottle({
                limit: 10,
                interval: 1010
            });
        throttled = throttle(i => {return {index: i, executed: Date.now()}});
    });

    it('throttles after delay', async function () {
        this.timeout(3000);

        const initPromises = [],
            promises = [];

        for (let i = 0; i < 11; i++) {
            initPromises.push(throttled(i));
        }

        await Promise.all(initPromises);

        await delay(50);

        const start = Date.now()

        for (let i = 0; i < 11; i++) {
            promises.push(throttled(i));
        }
        const results = await Promise.all(promises);

        results.forEach((result,index) => {
            console.log(result);
            if (index < 10) {
                (result.executed - start).should.be.within(0, 1010);
            } else {
                (result.executed - results[index - 10].executed).should.be.within(1000, 1020);
            }
        });
    });
});

function delay (ms) {
    return new Promise(resolve => { setTimeout(resolve, ms); });
}
```

### PR DESCRIPTION
**Pull request by schinkowitch**
Changes the contract of the throttle from never executing more than limit times within an interval to never waiting less than the interval between limit executions.

Fixes #26

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -21,6 +20,38 @@ test('main', async t => {
...
+test('limits after pause', async t => {
+	const limit = 10;
+	const interval = 1010;
+	const throttled = pThrottle({limit, interval})(() => Date.now());
+
+	const promises = [];
+
+	await throttled(0);
+
+	await new Promise(resolve => {
```
Comment (sindresorhus) [inline test.js:33]: Just use the `delay` dependency.

File: test.js [hunk 2]
```diff
@@ -21,6 +20,38 @@ test('main', async t => {
...
+
+	const start = Date.now();
+
+	for (let i = 0; i < 11; i++) {
+		promises.push(throttled(i));
+	}
+
+	const results = await Promise.all(promises);
+
+	results.forEach((executed, index) => {
```
Comment (sindresorhus) [inline test.js:45]: Use a `for-of` loop (just my preference).

File: index.js [hunk 3]
```diff
@@ -17,32 +17,24 @@ const pThrottle = ({limit, interval}) => {
...
 			let timeout;
 			return new Promise((resolve, reject) => {
 				const execute = () => {
+					ticks.pop();
 					resolve(function_.apply(this, args));
 					queue.delete(timeout);
 				};
 
 				const now = Date.now();
+				const executeAt = ticks.length < limit ? now : ticks[limit - 1] + interval;
```
Comment (sindresorhus) [inline index.js:33]:
```suggestion
				const runDate = ticks.length < limit ? now : ticks[limit - 1] + interval;
```

might be a clearer name.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Thanks for working on this fix? Would you be able to add some more tests? I'm a little bit paranoid about changing the implementation this much, so would be good to ensure corner cases are covered.

File: test.js (hunk 1 above)
Comment (Richienb) [reply test.js:33]: Use `yoctodelay` instead.

Comment (schinkowitch) [reply test.js]: Used delay

File: test.js (hunk 2 above)
Comment (schinkowitch) [reply test.js]: Ok

File: index.js (hunk 3 above)
Comment (schinkowitch) [reply index.js]: I changed the algorithm a bit so this is no longer relevant.

**Comment by schinkowitch** [comment]
I updated this pull request, so that the current algorithm is used by default. I added a strict option to use the strict throttling algorithm. I updated this algorithm compared to the previous commit, so that it only stores "limit" execution times instead of storing the execution times of all the queued calls, when using the strict algorithm.

**Comment by schinkowitch** [comment]
Merged changes and fixed merge error.

**Comment by schinkowitch** [comment]
Hi again.

I made a new package schinkowitch/strict-throttle that takes the ideas of the changes in this pull request and implements them with some improvements. So, feel free to close this without merging, but the code is yours if you wish to use it.

**Comment by sindresorhus** [comment]
Thanks :)

Sorry about the slow response. I just had too many PRs to review.
